tensorflow/compiler/xla/service/gpu/ir_emitter.cc

```diff
@@ -371,12 +364,21 @@ Status IrEmitter::EmitAtomicOperationUsingCAS(const HloComputation& computation,
     llvm::Type* address_int_type =
         module_->getDataLayout().getIntPtrType(output_address_type);
     atomic_memory_address = PtrToInt(output_address, address_int_type);
-    llvm::Value* mask = llvm::ConstantInt::get(address_int_type, -4);
+    llvm::Value* mask = llvm::ConstantInt::get(address_int_type, 3);
+    llvm::Value* offset = And(atomic_memory_address, mask);
+    mask = llvm::ConstantInt::get(address_int_type, -4);
     atomic_memory_address = And(atomic_memory_address, mask);
     atomic_memory_address =
         IntToPtr(atomic_memory_address, atomic_address_type);
-    binop_output_address = PointerBitCastOrAddrSpaceCast(cas_new_output_address,
-                                                         element_address_type);
+    binop_output_address =
+        Add(PtrToInt(cas_new_output_address, address_int_type), offset);
+    llvm::Type* cas_new_output_pointer_type = cas_new_output_address->getType();
+    CHECK_EQ(true, cas_new_output_pointer_type->isPointerTy());
+    unsigned correct_address_space =
+        cas_new_output_pointer_type->getPointerAddressSpace();
+    binop_output_address =
+        IntToPtr(binop_output_address,
+                 llvm::PointerType::get(element_type, correct_address_space));
```

Comment:
How about using  cas_new_output_pointer_type->getPointerAddressSpace directly ?

Reply:
Also something seems strange is binop_output_address out of 32 bit path has address space 0 ....although it works..

Reply:
in the "< 32bit" bit path, the problem is with the "cas_new_output".
we allocate it in "addrspace(5)", convert the pointer to int, potentially add and offset to it, and then convert it back to a pointer. The conversion back to pointer did not have the correct address space.

in the ">=32bit" path, we there is no pointer-->int-->pointer conversion, and hence we do not run into this scenario

Reply:
introduced the `correct_address_space` variable just to make the code more readable. would like to keep it that way, i fyou guys do not mind.

Reply:
You probably answered what I had in mind. Just to make sure ..I meant is there is disconnect in these address spaces 

< 32 bit (address space is 5)
binop_output_address =
        IntToPtr(binop_output_address,
                 llvm::PointerType::get(element_type, correct_address_space));
> 32  (address space is 0)
    binop_output_address = PointerBitCastOrAddrSpaceCast(cas_new_output_address,
                                                         element_address_type);

Reply:
I maybe wrong, but I thihk address-space is 5 in both cases.

cas_new_output is allocated on the stack (irrespective of element size) and hence address space for its pointer is 5. My assumption is that the `PointerBitCastOrAddrSpaceCast` routine does the right thing in maintaining the correct address space.

Reply:
Let me quickly verify ...

Reply:
binop_output_address is in 5 in < 32 path and 0 in 32 bit path

Reply:
interesting...then how does this work?

Reply:
Dont know.....we probably somehow get away ?  But address space 5 should be right.

I just tried address space 5 for 32 bit path, and it still works. Do you want to add that change as well?

Reply:
I could be convinced either way.  Lets see the result of the `rocm-xla` CI run (once it runs)...if it fails because of this issue, and then we know we need this change. If not, lets leave it as is

Reply:
CI run wont show any issue since 32 bit already passes. There is also a comment that says binop* should point to scratch 
" // binop_output_address points to the scratch memory that stores the
  // result of the binary operation.

Reply:
Thank you @sumesh13 for the offline discussion on this.

The address space is indeed 0 for the >=32bit case. I will push out a change soon to update the address space to 5 for the >=32bit case